### ISSUE TITLE
Search provider fails with the libretranslate backend

### ISSUE THREAD

**Issue by LyesSaadi** [dialect-app/dialect#188]
Well, that was a fun bug to debug :D!

# TLDR

When the libretranslate backend is selected, `search_provider` uses an obsolete dconf key, which causes the search provider to not work and to fail on startup.

This is due to this line of code:
https://github.com/dialect-app/dialect/blob/da1811a4638143d7f7e23f60fe6be549c479e9b0/dialect/search_provider/search_provider.in#L194

# The story

I have received some time ago [a puzzling bug from a Fedora user](https://bugzilla.redhat.com/show_bug.cgi?id=1961802) (I wasn't able to look into it sooner because of a lack of free time).

Here is the backtrace:
```
Truncated backtrace:
_exceptions.py:343:map_exceptions:httpx.LocalProtocolError: Missing hostname in URL.

Traceback (most recent call last):
  File "/usr/lib/python3.9/site-packages/httpx/_exceptions.py", line 326, in map_exceptions
    yield
  File "/usr/lib/python3.9/site-packages/httpx/_client.py", line 861, in _send_single_request
    (status_code, headers, stream, ext) = transport.request(
  File "/usr/lib/python3.9/site-packages/httpcore/_sync/connection_pool.py", line 193, in request
    raise LocalProtocolError("Missing hostname in URL.")
httpcore.LocalProtocolError: Missing hostname in URL.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/share/dialect/search_provider", line 208, in <module>
    TranslateService()
  File "/usr/share/dialect/search_provider", line 43, in __init__
    self.translator = self._get_translator()
  File "/usr/share/dialect/search_provider", line 193, in _get_translator
    translator = TRANSLATORS[backend](
  File "/usr/share/dialect/dialect/translators/libretrans.py", line 29, in __init__
    r = self.client.get(self.lang_url)
  File "/usr/lib/python3.9/site-packages/httpx/_client.py", line 907, in get
    return self.request(
  File "/usr/lib/python3.9/site-packages/httpx/_client.py", line 733, in request
    return self.send(
  File "/usr/lib/python3.9/site-packages/httpx/_client.py", line 767, in send
    response = self._send_handling_auth(
  File "/usr/lib/python3.9/site-packages/httpx/_client.py", line 805, in _send_handling_auth
    response = self._send_handling_redirects(
  File "/usr/lib/python3.9/site-packages/httpx/_client.py", line 837, in _send_handling_redirects
    response = self._send_single_request(request, timeout)
  File "/usr/lib/python3.9/site-packages/httpx/_client.py", line 861, in _send_single_request
    (status_code, headers, stream, ext) = transport.request(
  File "/usr/lib64/python3.9/contextlib.py", line 135, in __exit__
    self.gen.throw(type, value, traceback)
  File "/usr/lib/python3.9/site-packages/httpx/_exceptions.py", line 343, in map_exceptions
    raise mapped_exc(message, **kwargs) from exc  # type: ignore
httpx.LocalProtocolError: Missing hostname in URL.

Local variables in innermost frame:
mapping: {<class 'httpcore.TimeoutException'>: <class 'httpx.TimeoutException'>, <class 'httpcore.ConnectTimeout'>: <class 'httpx.ConnectTimeout'>, <class 'httpcore.ReadTimeout'>: <class 'httpx.ReadTimeout'>, <class 'httpcore.WriteTimeout'>: <class 'httpx.WriteTimeout'>, <class 'httpcore.PoolTimeout'>: <class 'httpx.PoolTimeout'>, <class 'httpcore.NetworkError'>: <class 'httpx.NetworkError'>, <class 'httpcore.ConnectError'>: <class 'httpx.ConnectError'>, <class 'httpcore.ReadError'>: <class 'httpx.ReadError'>, <class 'httpcore.WriteError'>: <class 'httpx.WriteError'>, <class 'httpcore.CloseError'>: <class 'httpx.CloseError'>, <class 'httpcore.ProxyError'>: <class 'httpx.ProxyError'>, <class 'httpcore.UnsupportedProtocol'>: <class 'httpx.UnsupportedProtocol'>, <class 'httpcore.ProtocolError'>: <class 'httpx.ProtocolError'>, <class 'httpcore.LocalProtocolError'>: <class 'httpx.LocalProtocolError'>, <class 'httpcore.RemoteProtocolError'>: <class 'httpx.RemoteProtocolError'>}
kwargs: {'request': <Request('GET', 'https:languages')>}
mapped_exc: <class 'httpx.LocalProtocolError'>
from_exc: <class 'httpcore.RemoteProtocolError'>
to_exc: <class 'httpx.RemoteProtocolError'>
message: 'Missing hostname in URL.'
```

You may say: Oh, wow, a httpx error? Well, yes, but actually no. As the error indicates, a URL request is done with a missing hostname. Intriguing isn't? This local variable `kwargs: {'request': <Request('GET', 'https:languages')>}` indicates that the request is done to `https:languages`. Yeah, this is definitely not a valid URL.

Looking more into the stack trace, it is an error happening in the libretranslate backend. And my first thought, is that they are using an exotic instance URL. So I launch dialect and I begin typing weird URLs, with ports and protocols, but alas I am unable to solve the issue. I think maybe this is an isolated issue with a particular instance, so I ask the reporter to give me what instance he is using, but [when checking the Bug Reporter](https://retrace.fedoraproject.org/faf/problems/bthash/?bth=b9f794d44a4d96e51a6e69d86036122ea3176c76), oh god, 966 automatic reports (the number is still rising as I am writing this :P)... This is definitely not normal (while only a subset of users activates automatic bug reporting)! At the same time, when looking into the function which accepts and stores the backend instance, it seems quite solid. And playing with dconf does nothing as well...

So, today, I keep looking into it, and something else catches my attention: This bug is coming from the search provider. That would absolutely explain why there are so much reports as each time the session is opened, it fails. But, since the issue is with the instance URL, I there's no way the issue is coming from there, right? Right (:?

I still take a look, just to follow the bug path and to understand the error as a whole, when I see this:
```python
    def _get_translator(self):
        backend = Settings.get().backend
        if TRANSLATORS[backend].supported_features['change-instance']:
            translator = TRANSLATORS[backend](
                base_url=Settings.get().get_string(f'{TRANSLATORS[backend].name}-instance')
            )
        else:
            translator = TRANSLATORS[backend]()
        return translator
```

What immediately catches my attention is the format string, `f'{TRANSLATORS[backend].name}-instance'`. I think, ok, it is searching for `libretranslate-instance`, legitimate. I decide to take a look at how this value is represented in dconf. When I find this:

```
DEPRECATED: This key is deprecated and normally ignored.
```

The `search_provider` script was not updated when the keys in dconf were changed! I then immediately launch `/usr/share/dialect/search_provider`, and bingo:

```
Traceback (most recent call last):
  File "/usr/share/dialect/search_provider", line 208, in <module>
    TranslateService()
  File "/usr/share/dialect/search_provider", line 43, in __init__
    self.translator = self._get_translator()
  File "/usr/share/dialect/search_provider", line 193, in _get_translator
    translator = TRANSLATORS[backend](
  File "/usr/share/dialect/dialect/translators/libretrans.py", line 29, in __init__
    r = self.client.get(self.lang_url)
  File "/var/home/lyes/.local/lib/python3.9/site-packages/httpx/_client.py", line 755, in get
    return self.request(
  File "/var/home/lyes/.local/lib/python3.9/site-packages/httpx/_client.py", line 590, in request
    request = self.build_request(
  File "/var/home/lyes/.local/lib/python3.9/site-packages/httpx/_client.py", line 194, in build_request
    url = self.merge_url(url)
  File "/var/home/lyes/.local/lib/python3.9/site-packages/httpx/_client.py", line 214, in merge_url
    url = self.base_url.join(relative_url=url)
  File "/var/home/lyes/.local/lib/python3.9/site-packages/httpx/_models.py", line 222, in join
    return URL(relative_url)
  File "/var/home/lyes/.local/lib/python3.9/site-packages/httpx/_models.py", line 92, in __init__
    raise InvalidURL("No host included in URL.")
httpx._exceptions.InvalidURL: No host included in URL.
```

Anyway, funny consequences of that, I now have detailed metrics of every time a Fedora user, which activated automatic bug reporting, installed dialect, chose the libretranslate backend & enabled the search provider, opened a new session :P! No idea what to do with that information XD!

**Comment by rafaelmardojai** [comment]
Thanks for the detailed report! 

I'm a bit embarrassed with this going through our noses, but I guess it happens in the best of families :P

**Comment by LyesSaadi** [comment]
Fedora update:
Fedora rawhide: https://bodhi.fedoraproject.org/updates/FEDORA-2021-3cbec36dfc
Fedora 34: https://bodhi.fedoraproject.org/updates/FEDORA-2021-c54b73dfb1
Fedora 33: https://bodhi.fedoraproject.org/updates/FEDORA-2021-38ecb2c2be
